modules/userdirectory-canvas/src/main/java/org/opencastproject/userdirectory/canvas/CanvasUserProviderInstance.java

```diff
@@ -0,0 +1,705 @@
+      pid, canvasUrl, courseIdentifierProperty, coursePattern, userPattern, userNameProperty, String.join(", ", instructorRoles), cacheSize, cacheExpiration);
+
+    // Setup the caches
+    cache = CacheBuilder.newBuilder().maximumSize(cacheSize).expireAfterWrite(cacheExpiration, TimeUnit.MINUTES)
+            .build(new CacheLoader<String, Object>() {
+              @Override
+              public Object load(String id) throws Exception {
+                User user = loadUserFromCanvas(id);
+                return user == null ? nullToken : user;
+              }
+            });
+
+    registerMBean(pid);
+  }
+
+  @Override
+  public String getName() {
+    return PROVIDER_NAME;
+  }
+
+  /**
+   * Registers an MXBean.
+   */
+  protected void registerMBean(String pid) {
+    // Register with jmx
+    requests = new AtomicLong();
+    canvasLoads = new AtomicLong();
+    try {
+      ObjectName name;
+      name = CanvasUserProviderFactory.getObjectName(pid);
+      Object mbean = this;
+      MBeanServer mbs = ManagementFactory.getPlatformMBeanServer();
+      try {
+        mbs.unregisterMBean(name);
+      } catch (InstanceNotFoundException e) {
+        logger.debug(name + " was not registered");
+      }
+      mbs.registerMBean(mbean, name);
+    } catch (Exception e) {
+      logger.error("Exception while registering {} as an mbean: {}", this, e);
+    }
+  }
+
+  // UserProvider methods
+
+  /**
+   * {@inheritDoc}
+   * 
+   * @see org.opencastproject.security.api.UserProvider#getOrganization()
+   */
+  @Override
+  public String getOrganization() {
+    return organization.getId();
+  }
+
+  /**
+   * {@inheritDoc}
+   * 
+   * @see org.opencastproject.security.api.UserProvider#loadUser(java.lang.String)
+   */
+  @Override
+  public User loadUser(String userId) {
+    logger.debug("loadUser({})", userId);
+    requests.incrementAndGet();
+    try {
+      Object user = cache.getUnchecked(userId);
+      if (user == nullToken) {
+        logger.debug("Returning null user from cache");
+        return null;
+      } else {
+        logger.debug("Returning user " + userId + " from cache");
+        return (JaxbUser) user;
+      }
+    } catch (ExecutionError e) {
+      logger.warn("Exception while loading user {}: {}", userId, e);
+      return null;
+    } catch (UncheckedExecutionException e) {
+      logger.warn("Exception while loading user {}: {}", userId, e);
+      return null;
+    }
+  }
+
+  /**
+   * Loads a user from Canvas.
+   * 
+   * @param userId
+   * @return The user
+   */
+  protected User loadUserFromCanvas(String userId) {
+    if (cache == null) {
+      throw new IllegalStateException("The Canvas user detail service has not yet been configured");
+    }
+
+    // Don't answer for admin, anonymous or empty user
+    if ("admin".equals(userId) || "".equals(userId) || "anonymous".equals(userId)) {
```

Comment:
This I'm not too sure on.  It seems to be copied from the [moodle user provider](https://github.com/opencast/opencast/blob/f5fd263fb90f7eea0c984bf9a63d6cb87c6e297f/modules/userdirectory-moodle/src/main/java/org/opencastproject/userdirectory/moodle/MoodleUserProviderInstance.java#L477)

What would the configuration be if I needed to compare this to it?

Reply:
Hm, seems like the first LMS user provided did it wrong and everyone has since just copied that code. Great -_-". Well, at least we found a bug. I'll try to provide a reference implementation. Hopefully tomorrow.

Reply:
As promised, I created a reference implementation by fixing the Moodle user provider in pull request #902